### PR TITLE
Assignment 3 PR Team F

### PR DESCRIPTION
**Pull request by noahpy**
This is an alternate PR to the original assignment3 PR, which was needed to resolve merge conflicts.
To ensure good practice, branch protection rules were added to this branch too.

### CODE REVIEW

**Comment by Chryzl** [comment]
**Bug Fix**
- The bug we reported in our report has been fixed shortly after the deadline
- The issue was with our CellGrid Algorithm (accelerating particles in bottom-left-back corner)
- As the discovery was after the deadline, we did not want to merge the fix without telling you about it to guarantee fair grading
- Please see the PR that would merge the fix into this branch -> #41 

- When you run our program, feel free to also check out the fix

We apologize for the inconvenience

File: bench/results/lj_big_plot_log.png
Comment (manishmishra6016) [inline bench/results/lj_big_plot_log.png:1]: In general, I would recommend to provide a python script that can generate these plots rather than the plots themselves. One reason is the size of file. But this is also fine if you prefer this, I have no strong feelings about this.

File: cmake/modules/xsd.cmake [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# Check if xsdcxx is installed
+find_program(XSD_EXECUTABLE NAMES xsdcxx)
+
+# Dont do anything if xsdcxx is not installed
+if (XSD_EXECUTABLE)
+    file(GLOB_RECURSE ALL_XSD
+        "${CMAKE_CURRENT_SOURCE_DIR}/src/io/xsd/*.xsd"
+    )
+    add_custom_target(xsd
+        COMMAND ${XSD_EXECUTABLE} cxx-tree --hxx-suffix=.h --cxx-suffix=.cpp --std c++11 --generate-doxygen --generate-serialization ${ALL_XSD}
```
Comment (manishmishra6016) [inline cmake/modules/xsd.cmake]: I guess a recommendation would be use a more recent C++, maybe C++ 17. Does it cause problems for you? 

You can also enforce the standard throughout the code in `CMakeLists.txt` (it is a nice practice to enforce a standard).

File: README.md [hunk 2]
```diff
@@ -53,6 +53,16 @@ src/MolSim ../input/eingabe-sonne.txt -e 1000
...
 src/MolSim ../input/clusters.txt -c -s 1 -d 0.0002 -e 5
 ```
+#### Assignment 3 simulation
+Cuboid collision:
+```
+src/MolSim ../input/assign3.xml -x -s 3
+```
+Falling drop:
+```
+src/MolSim ../input/falling_drop.xml -x -s 3
```
Comment (manishmishra6016) [inline README.md:63]: I would explain the tag `-s` and `-x` to a user. I understand `-s` is for the simulation type (as I can see in help) but it doesn't tell me what type of simulation and what are the other option, so maybe a table here would be nice?

File: .molsim.1 [hunk 3]
```diff
@@ -61,6 +68,12 @@ Specify simulation type (default: 0).
 .br
 1 = LennardJonesSimulation
 .br
+.br
+2 = LennardJonesSimulation with Linked Cells
+.br
+.br
+3 = LennardJonesSimulation with Linked Cells + Boundary Conditions
```
Comment (manishmishra6016) [inline .molsim.1:75]: It would be nice to have two inputs (container: directSum, LinkedCells, and forceFunctor: LJ, gravity, etc.). This way we will have more flexibility. For eg. right now, I cannot run planet simulation with LinkedCells! While it is not supercritical here, it would be nice to restructure the code to make this possible (also for new algorithms one may add in future).

File: docs/report/report3/res/sphere.png
Comment (manishmishra6016) [inline docs/report/report3/res/sphere.png:1]: you could have a shared folder for images, and you can use them for both report and presentation to avoid repetition.

File: input/assign3.xml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      <z>0</z>
+    </domainOrigin>
+    <domainSize>
+      <x>180</x>
+      <y>90</y>
+      <z>0</z>
+    </domainSize>
+    <cutoff>3.0</cutoff>
+    <boundaries>
+      <bound_four>overflow</bound_four>
```
Comment (manishmishra6016) [inline input/assign3.xml]: I would change the name `overflow` to `outflow` everywhere, just to stick to standard terminology.

File: src/io/argparse/argparse.cpp [hunk 5]
```diff
@@ -102,17 +105,37 @@ void convertToDouble(char* optarg, double& value)
...
-    char* argsv[],
-    double& end_time,
-    double& delta_t,
-    double& epsilon,
-    double& sigma,
-    std::string& input,
-    unsigned& reader_type,
-    unsigned& writer_type,
-    unsigned& simulation_type)
+SimulationType unsignedToSimulationType(unsigned value)
```
Comment (manishmishra6016) [inline src/io/argparse/argparse.cpp:108]: refer to previous comment on how to construct simulation from container and forceFunctor. Here I would just keep two options - planet, molecular.

File: src/io/fileWriter/writerFactory.cpp [hunk 6]
```diff
@@ -4,15 +4,17 @@
...
     switch (type) {
-    case 0:
-        return std::make_unique<outputWriter::VTKWriter>();
-    case 1:
-        return std::make_unique<outputWriter::XYZWriter>();
+    case WriterType::VTK:
+        spdlog::info("Initializing VTKWriter...");
+        return std::make_unique<outputWriter::VTKWriter>(out_name);
+    case WriterType::XYZ:
+        spdlog::info("Initializing XYZWriter...");
```
Comment (manishmishra6016) [inline src/io/fileWriter/writerFactory.cpp:14]: nice use of logging!

File: src/io/xsd/simulation.cpp [hunk 7]
```diff
@@ -0,0 +1,2717 @@
...
+{
+    return this->z_.get();
+}
+
+intVec_t::z_type& intVec_t::z()
+{
+    return this->z_.get();
+}
+
+void intVec_t::z(const z_type& x)
```
Comment (manishmishra6016) [inline src/io/xsd/simulation.cpp]: a minor comment on styling here, stick to either `snake_case` or `camelBack`

File: src/models/Particle.cpp [hunk 8]
```diff
@@ -98,6 +98,11 @@ void Particle::setX(const std::array<double, 3>& x_new)
     x = x_new;
 }
 
+void Particle::setM(double m_new)
```
Comment (manishmishra6016) [inline src/models/Particle.cpp:101]: generally I would try to handle setting constants like mass in constructors only. It is not safe to allow users to change the mass of a particle. So, this might be something to look into if and when you refactor your code.

File: src/models/ParticleContainer.cpp [hunk 9]
```diff
@@ -21,6 +22,20 @@ void ParticleContainer::addParticle(const Particle& p)
...
+    // remove_if rearranged the elements in the vector, so we need to erase the elements after the
+    // last valid one
+    particles.erase(
+        std::remove_if(
+            particles.begin(),
+            particles.end(),
+            [&particleMap](const Particle& p) {
+                return particleMap.find(const_cast<Particle*>(&p)) != particleMap.end();
+            }),
+        particles.end());
```
Comment (manishmishra6016) [inline src/models/ParticleContainer.cpp:36]: While removing is fine, I am not sure about how it affects the performance. In general, it is a nice strategy to reuse the particles you create. But for the worksheet, this is fine.

Why did you need `const_cast`? Just curious here.

File: src/models/generators/SphereParticleCluster.cpp [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+            double capRadiusOffset = discRadius == 1 ? spacing * 0.25 : 0;
+            double realRadius = getRealRadius(sphereRadius, z_offset - capRadiusOffset);
+            totalNumberOfParticles += 2 * getNumberOfParticlesDisc(realRadius);
+            z_offset += spacing;
+        }
+}
+
+size_t SphereParticleCluster::getTotalNumberOfParticles() const
+{
+    return totalNumberOfParticles;
```
Comment (manishmishra6016) [inline src/models/generators/SphereParticleCluster.cpp:47]: this should be initialized in constructor, so you can access it immediately after the creating the sphere.

File: src/models/linked_cell/CellGrid.h [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+    /// A vector storing the indices of halo cells.
+    std::vector<CellIndex> haloCells;
+
+    /* ##### Detailed Iterator Definitions ##### */
+public:
+    /** @class BoundaryIterator
+     * @brief Iterator for iterating over boundary cells.
+     */
+    class BoundaryIterator {
```
Comment (manishmishra6016) [inline src/models/linked_cell/CellGrid.h:157]: I would personally avoid nested classes as they make header file longer and less readable in my opinion. But it is fine if you prefer it this way.

File: src/physics/boundaryConditions/OverflowBoundary.cpp [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        particle.get().setF({ 0, 0, 0 });
+        particle.get().setOldF({ 0, 0, 0 });
+    }
+
+    /*
+     * Problem:
+     * If we remove the particles form the particle containter, the references in the cell will be
+     * invalid thus the iterator will be invalid that is used in the force calculations.
+     * We need to change the way the particle is deleted, or we need to change the way, the
+     * grid is updated
```
Comment (manishmishra6016) [inline src/physics/boundaryConditions/OverflowBoundary.cpp:39]: the solution would be to use particle pointers or just tag the particles and ignore them. And even better would be to try and reuse these tagged particles if possible.

File: src/physics/boundaryConditions/SoftReflectiveBoundary.cpp [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    std::array<double, 3> normal = getNormalVectorOfBoundary(position);
+    for (auto boundaryCellIndex : LGDSim.getGrid().boundaryCellIterator(position)) {
+        for (auto& particle :
+             LGDSim.getGrid()
+                 .cells[boundaryCellIndex[0]][boundaryCellIndex[1]][boundaryCellIndex[2]]
+                 ->getParticles()) {
+            std::array<double, 3> pos = particle.get().getX();
+            std::array<double, 3> pointOnBoundaryPlane = getPointOnBoundaryPlane(LGDSim);
+
+            std::array<double, 3> diff = pointOnBoundaryPlane - pos;
```
Comment (manishmishra6016) [inline src/physics/boundaryConditions/SoftReflectiveBoundary.cpp:39]: Missing check: this diff must be smaller than `1.12 · σ`, only then we reflect the particle.

File: src/physics/forceCal/forceCal.cpp [hunk 14]
```diff
@@ -62,19 +78,57 @@ void force_lennard_jones(const Simulation& sim)
...
+                if (doLoopFor2D) {
+                    doLoopFor2D = false; // only do it once
+                    z = 0;
+                }
+
+                // calculate the LJ forces in the cell
+                for (auto it = cellGrid.cells.at(x).at(y).at(z)->beginPairs();
+                     it != cellGrid.cells.at(x).at(y).at(z)->endPairs();
+                     ++it) {
+                    auto pair = *it;
```
Comment (manishmishra6016) [inline src/physics/forceCal/forceCal.cpp:122]: Check missing: particles may not be within cutoff distance.

File: src/simulation/LennardJonesDomainSimulation.h [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+#include "simulation/linkedLennardJonesSim.h"
+
+/**
+ * @brief Simulation class for the Lennard-Jones simulation within a restricted domain and linked
+ * cell algorithm
+ * @details This class is a subclass of the Lennard-Jones simulation class using linked cells and is
+ * used to simulate the movement of particles according to the Lennard-Jones potential within a
+ * restricted domain. This is the simulation class used in the third problem sheet.
+ */
+class LennardJonesDomainSimulation : public LinkedLennardJonesSimulation {
```
Comment (manishmishra6016) [inline src/simulation/LennardJonesDomainSimulation.h:12]: I would recommend restructuring a bit so that simulation class is not disturbed (or you don't need to create a bunch of new ones) and instead just create new containers and functors that can be used in the simulation. This would be extendable in long run.

File: tests/models/generators/testSphereParticleGenerator.cpp [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+        dimensions);
+    generator.registerCluster(std::make_unique<SphereParticleCluster>(cluster));
+    generator.generateClusters();
+
+    for (auto& p : container) {
+        double dist = ArrayUtils::L2Norm(p.getX() - origin);
+        EXPECT_LE(dist, radius * spacing);
+    }
+}
+/*
```
Comment (manishmishra6016) [inline tests/models/generators/testSphereParticleGenerator.cpp]: why is this commented out? I liked the idea of the test but if you don't want to keep it, you can just remove it. do not keep dead code in master branch.

**Review by manishmishra6016 — COMMENTED** [review]
**Note**
_This is only feedback regarding the code directly. Grade feedback will be on Moodle once we are through with all groups._

-----------------------------------------------------------------------------------------------------

Thanks for the nice submission :muscle: 
You were able to complete most objectives. I would however recommend looking at refactoring the code in a way that you can have more freedom in choosing combination of simulation type, container, and force functor. 

More importantly, Linked Cells still needs to check for the cutoff distance which is presently missing, and similarly reflective boundary is not applied to all the particles in boundary cells. Doxygen main page is still missing. These are rather important.

Have a look at the comments and see if they make sense. Keep up your nice efforts and I am sure you will do great in next worksheets as well :smile:

File: src/io/fileWriter/writerFactory.cpp (hunk 6 above)
Comment (noahpy) [reply src/io/fileWriter/writerFactory.cpp:14]: Thanks :)

File: src/models/Particle.cpp (hunk 8 above)
Comment (noahpy) [reply src/models/Particle.cpp:101]: Yes this is true, but when we want to "reuse" Particles, it is convinient to be able to modify the existing Particle object :)

File: src/models/generators/SphereParticleCluster.cpp (hunk 10 above)
Comment (noahpy) [reply src/models/generators/SphereParticleCluster.cpp:47]: We don't fully understand what should be initialized. If you mean `totalNumberOfParticles`, it is initialized in line 34 :)

File: src/models/linked_cell/CellGrid.h (hunk 11 above)
Comment (noahpy) [reply src/models/linked_cell/CellGrid.h:157]: Okay, we will look out on the next iterations.
For now, I think we will leave it at this.

File: src/physics/boundaryConditions/OverflowBoundary.cpp (hunk 12 above)
Comment (noahpy) [reply src/physics/boundaryConditions/OverflowBoundary.cpp:39]: Yes, we are tackling this in this cycle :)

File: tests/models/generators/testSphereParticleGenerator.cpp (hunk 16 above)
Comment (noahpy) [reply tests/models/generators/testSphereParticleGenerator.cpp]: I think we would delete these :)

File: src/physics/boundaryConditions/SoftReflectiveBoundary.cpp (hunk 13 above)
Comment (Chryzl) [reply src/physics/boundaryConditions/SoftReflectiveBoundary.cpp:39]: We have changed it in the code (see PR-fix).
However, we need to generalize it once we introduce multiple particle types with different sigmas. We will refactor the Assignment 4 branch and introduce the change once we know our code structure for the 4th assignment.

File: src/models/generators/SphereParticleCluster.cpp (hunk 10 above)
Comment (manishmishra6016) [reply src/models/generators/SphereParticleCluster.cpp:47]: Yes, I missed it. All good here :)

File: src/models/Particle.cpp (hunk 8 above)
Comment (manishmishra6016) [reply src/models/Particle.cpp:101]: I understand and reusing is really nice. You could think of using copy constructor if it makes sense for your case. But otherwise this is fine :)

File: bench/results/lj_big_plot_log.png
Comment (noahpy) [reply bench/results/lj_big_plot_log.png:1]: I think that would be good practice. I added the git repo I used for generating the plots as a submodule to the project :)

File: cmake/modules/xsd.cmake (hunk 1 above)
Comment (noahpy) [reply cmake/modules/xsd.cmake]: That is true :) I did implement the change in both CMakeFiles.txt and the xsd module.

File: README.md (hunk 2 above)
Comment (noahpy) [reply README.md:63]: I thought of listing the meaning of the options, but thought this would get too long for a help message. 
As specified in the help message, one can read all the details in the man page. 
I think this suffices for now :)

File: input/assign3.xml (hunk 4 above)
Comment (noahpy) [reply input/assign3.xml]: Oh yes, that is indeed a mistake!

File: src/io/xsd/simulation.cpp (hunk 7 above)
Comment (noahpy) [reply src/io/xsd/simulation.cpp]: I think the _t here is more of a suffix to denote that this is a type rather than indicating a snake case name. I think using snake case for the whole name would make the suffix less visible, but you are right that this styling is not that consistent. I will look out for this for future namings, leaving this for now so that we do not need to rewrite a lot of input files. Hope this is ok!

File: src/models/ParticleContainer.cpp (hunk 9 above)
Comment (noahpy) [reply src/models/ParticleContainer.cpp:36]: Great point, we are implementing the whole reusing in the next worksheet :)
We are not sure why we used `const_cast`, maybe a recommendation by clang-tidy?

**Comment by noahpy** [comment]
@manishmishra6016 Thank you for your thourough review and comments! :)
We have read and addressed every comment on a seperate branch pr-fixes, which we are going to merge in this branch, as soon as it finishes up.
The only thing we have not addressed yet is the restructuring of the simulation class with functors for more flexibility. While this would be very nice and handy, we would like to first finish the current worksheet, as this refactoring is quite some work :) Would that be fine for you? 
I will re-request as soon as we merged the other fixes!

File: docs/report/report3/res/sphere.png
Comment (noahpy) [reply docs/report/report3/res/sphere.png:1]: Yes, but I think we will leave like this for now. We may restructure if there are considerably many duplicate images :)

**Comment by manishmishra6016** [comment]
> The only thing we have not addressed yet is the restructuring of the simulation class with functors for more flexibility. While this would be very nice and handy, we would like to first finish the current worksheet, as this refactoring is quite some work :) Would that be fine for you? I will re-request as soon as we merged the other fixes!

Hi, thank you for all these changes. I would suggest the same. Focus on implementing the tasks first and then restructuring if you still have time. Some of these comments are just to inform you on what might be different but we don't necessarily expect you to be able to implement all the changes. We understand that tasks themselves are quite time consuming :) good luck with the worksheet :) 

Also, feel free to merge as you wish after the reviews in general.

**Review by Chryzl — COMMENTED** [review]
We have implemented the requested changes and merged them into this branch. We will go ahead and merge the branch

**Review by Chryzl — APPROVED** [review]
Merging...
